cloudflare/resource_cloudflare_spectrum_app.go

```diff
@@ -0,0 +1,313 @@
+package cloudflare
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"time"
+
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"github.com/pkg/errors"
+)
+
+const (
+	cloudflareOn  = "on"
+	cloudflareOff = "off"
+)
+
+func resourceCloudflareSpectrumApp() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudflareSpectrumAppCreate,
+		Read:   resourceCloudflareSpectrumAppRead,
+		Update: resourceCloudflareSpectrumAppUpdate,
+		Delete: resourceCloudflareSpectrumAppDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceCloudflareSpectrumAppImport,
+		},
+
+		SchemaVersion: 0,
+		Schema: map[string]*schema.Schema{
+			"zone_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"protocol": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"dns": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"type": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+
+			"origin_direct": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+
+			"origin_dns": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+
+			"origin_port": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+
+			"tls": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  cloudflareOff,
+				ValidateFunc: validation.StringInSlice([]string{
+					cloudflareOn,
+					cloudflareOff,
+				}, false),
+			},
+
+			"ip_firewall": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+
+			"proxy_protocol": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"created_on": {
```

Comment:
I've generally left the `{Created,Modified}On` values out of Terraform as they cannot be updated or managed so it seemed a little redundant.